File: VB/SnapModule/Helpers/SnapDocumentHelper.vb
Imports System
Imports System.Collections.Generic
Imports DevExpress.ExpressApp
Imports SnapModule.Base
Imports DevExpress.Snap
Imports DevExpress.XtraPrinting
Imports System.IO
Imports DevExpress.Snap.Core.API
Imports DevExpress.Data.Filtering
Imports DevExpress.ExpressApp.Editors
Imports System.Data
Imports System.Collections
Imports DevExpress.ExpressApp.SystemModule
Imports SnapModule.Editors

Namespace SnapModule.Helpers
    Public Class SnapDocumentHelper
        Private documentData As IDocumentData
        Private objectSpace As IObjectSpace
        Private application As XafApplication
        Private inPlaceCriteria As CriteriaOperator
        Private document As SnapDocument

        Public Sub New(ByVal documentData As IDocumentData, ByVal objectSpace As IObjectSpace, ByVal application As XafApplication)
            Me.New(documentData, objectSpace, application, Nothing)
        End Sub

        Public Sub New(ByVal documentData As IDocumentData, ByVal objectSpace As IObjectSpace, ByVal application As XafApplication, ByVal inPlaceCriteria As CriteriaOperator)
            Me.documentData = documentData
            Me.objectSpace = objectSpace
            Me.application = application
            Me.inPlaceCriteria = inPlaceCriteria
        End Sub

        Public Shared Sub ShowDesigner(ByVal documentData As IDocumentData, ByVal objectSpace As IObjectSpace, ByVal application As XafApplication)
            Dim helper As New SnapDocumentHelper(documentData, objectSpace, application)
            Dim designer As New SnapDesigner(helper)
            designer.ShowDialog()
        End Sub
        Public Shared Sub ShowPrintPreview(ByVal documentData As IDocumentData, ByVal objectSpace As IObjectSpace)
            ShowPrintPreview(documentData, objectSpace, Nothing)
        End Sub
        Public Shared Sub ShowPrintPreview(ByVal documentData As IDocumentData, ByVal objectSpace As IObjectSpace, ByVal inPlaceCriteria As CriteriaOperator)
            Using server As New SnapDocumentServer()
                Dim helper As New SnapDocumentHelper(documentData, objectSpace, Nothing, inPlaceCriteria)
                helper.LoadDocument(server.Document)
                ' Mail Merge
                If server.Document.DataSources.Count = 1 AndAlso server.Options.SnapMailMergeVisualOptions.DataSource Is server.Document.DataSources(0).DataSource Then
                    Using stream As New MemoryStream()
                        server.SnapMailMerge(stream, SnapDocumentFormat.Snap)
                        stream.Flush()
                        stream.Seek(0, SeekOrigin.Begin)
                        server.LoadDocument(stream, SnapDocumentFormat.Snap)
                    End Using
                End If
                Using ps As New PrintingSystem()
                    Dim link As New PrintableComponentLink(ps)
                    link.Component = server
                    link.CreateDocument()
                    link.ShowPreviewDialog()
                End Using
            End Using
        End Sub
        Public Sub LoadDocument(ByVal document As SnapDocument)
            Me.document = document
            If documentData.Content IsNot Nothing Then
                Using stream As New MemoryStream()
                    stream.Write(documentData.Content, 0, documentData.Content.Length)
                    stream.Flush()
                    stream.Seek(0, SeekOrigin.Begin)
                    document.LoadDocument(stream, SnapDocumentFormat.Snap)
                End Using
            End If
            SynchronizeDataSources()
        End Sub

        Private Sub SynchronizeDataSources()
            Dim dataSourcesToRemove As New List(Of DataSourceInfo)()
            For Each dataSourceInfo As DataSourceInfo In document.DataSources
                If Not(TypeOf dataSourceInfo.DataSource Is DataSet) Then
                    Dim exists As Boolean = False
                    For Each dataSource As IDocumentDataSource In documentData.DataSources
                        If dataSource.Name = dataSourceInfo.DataSourceName Then
                            exists = True
                            Exit For
                        End If
                    Next dataSource
                    If Not exists Then
                        dataSourcesToRemove.Add(dataSourceInfo)
                    End If
                End If
            Next dataSourceInfo
            For Each dataSourceInfo As DataSourceInfo In dataSourcesToRemove
                document.DataSources.Remove(dataSourceInfo)
            Next dataSourceInfo
            For Each dataSource As IDocumentDataSource In documentData.DataSources
                AddDataSource(dataSource, dataSource Is documentData.InplaceDocumentDataSource)
            Next dataSource
            document.Fields.Update()
        End Sub
        Private Sub AddDataSource(ByVal dataSource As IDocumentDataSource, ByVal isInplaceDataSource As Boolean)
            Dim criteria As CriteriaOperator = If(isInplaceDataSource, inPlaceCriteria, Nothing)
            If Not String.IsNullOrEmpty(dataSource.Criteria) Then
                Dim dataSourceCriteria As CriteriaOperator = CriteriaEditorHelper.GetCriteriaOperator(dataSource.Criteria, dataSource.DataType, objectSpace)
                If ReferenceEquals(criteria, Nothing) Then
                    criteria = dataSourceCriteria
                Else
                    criteria = New GroupOperator(GroupOperatorType.And, criteria, dataSourceCriteria)
                End If
            End If
            Dim collectionSource As New CollectionSource(objectSpace, dataSource.DataType, CollectionSourceDataAccessMode.Client, CollectionSourceMode.Proxy)
            collectionSource.Criteria("SnapDataSourceFiler") = criteria
            If document.DataSources(dataSource.Name) IsNot Nothing Then
                document.DataSources(dataSource.Name).DataSource = collectionSource.Collection
            Else
                document.DataSources.Add(dataSource.Name, collectionSource.Collection)
            End If
        End Sub
        Friend Sub SaveDocument()
            If objectSpace.IsNewObject(documentData) Then
                ShowDocumentDataObject()
            End If
            Using stream As New MemoryStream()
                document.SaveDocument(stream)
                stream.Seek(0, SeekOrigin.Begin)
                Dim bytes(stream.Length - 1) As Byte
                stream.Read(bytes, 0, bytes.Length)
                documentData.Content = bytes
                objectSpace.CommitChanges()
            End Using
        End Sub
        Friend Sub ShowDocumentDataObject()
            Dim documentDetailView As DetailView = application.CreateDetailView(objectSpace, documentData, False)
            documentDetailView.Caption = "Save Document"
            Dim svp As New ShowViewParameters(documentDetailView)
            svp.TargetWindow = TargetWindow.NewModalWindow
            svp.Context = TemplateContext.PopupWindow
            Dim dc As DialogController = application.CreateController(Of DialogController)()
            dc.SaveOnAccept = False
            svp.Controllers.Add(dc)
            application.ShowViewStrategy.ShowView(svp, New ShowViewSource(Nothing, Nothing))
            SynchronizeDataSources()
        End Sub
        Friend Sub CreateNewDocument(ByVal document As SnapDocument)
            documentData = DirectCast(objectSpace.CreateObject(DirectCast(documentData, Object).GetType()), IDocumentData)
        End Sub
        Friend Sub ShowNewDataSourceDialog()
            Dim newDataSourceObjectSpace As IObjectSpace = objectSpace.CreateNestedObjectSpace()
            Dim dataSourceType As Type = XafTypesInfo.Instance.FindTypeInfo(DirectCast(documentData, Object).GetType()).FindMember("DataSources").ListElementType
            Dim dataSource As IDocumentDataSource = DirectCast(newDataSourceObjectSpace.CreateObject(dataSourceType), IDocumentDataSource)
            Dim dataSourceDetailView As DetailView = application.CreateDetailView(newDataSourceObjectSpace, dataSource)
            Dim svp As New ShowViewParameters(dataSourceDetailView)
            svp.TargetWindow = TargetWindow.NewModalWindow
            svp.Context = TemplateContext.PopupWindow
            Dim dc As DialogController = application.CreateController(Of DialogController)()
            svp.Controllers.Add(dc)
            AddHandler dc.Accepting, Sub()
                newDataSourceObjectSpace.CommitChanges()
                Dim newDataSource As IDocumentDataSource = DirectCast(objectSpace.GetObject(dataSource), IDocumentDataSource)
                documentData.DataSources.Add(newDataSource)
                AddDataSource(newDataSource, False)
            End Sub
            application.ShowViewStrategy.ShowView(svp, New ShowViewSource(Nothing, Nothing))
        End Sub
        Friend Function QueryCanClose() As Boolean
            Dim result As ConfirmationResult = application.AskConfirmation(ConfirmationType.NeedSaveChanges)
            If result = ConfirmationResult.Cancel Then
                Return False
            End If
            If result = ConfirmationResult.Yes Then
                SaveDocument()
            End If
            Return True
        End Function
        Friend Function GetDocumentName() As String
            Return documentData.DocumentName
        End Function
        Friend ReadOnly Property Modified() As Boolean
            Get
                Return objectSpace.IsModified
            End Get
        End Property
    End Class
End Namespace

File: CS/SnapModule/Helpers/SnapDocumentHelper.cs
using System;
using System.Collections.Generic;
using DevExpress.ExpressApp;
using SnapModule.Base;
using DevExpress.Snap;
using DevExpress.XtraPrinting;
using System.IO;
using DevExpress.Snap.Core.API;
using DevExpress.Data.Filtering;
using DevExpress.ExpressApp.Editors;
using System.Data;
using System.Collections;
using DevExpress.ExpressApp.SystemModule;
using SnapModule.Editors;

namespace SnapModule.Helpers {
    public class SnapDocumentHelper {
        private IDocumentData documentData;
        private IObjectSpace objectSpace;
        private XafApplication application;
        private CriteriaOperator inPlaceCriteria;
        private SnapDocument document;

        public SnapDocumentHelper(IDocumentData documentData, IObjectSpace objectSpace, XafApplication application)
            : this(documentData, objectSpace, application, null) { }

        public SnapDocumentHelper(IDocumentData documentData, IObjectSpace objectSpace, XafApplication application, CriteriaOperator inPlaceCriteria) {
            this.documentData = documentData;
            this.objectSpace = objectSpace;
            this.application = application;
            this.inPlaceCriteria = inPlaceCriteria;
        }

        public static void ShowDesigner(IDocumentData documentData, IObjectSpace objectSpace, XafApplication application) {
            SnapDocumentHelper helper = new SnapDocumentHelper(documentData, objectSpace, application);
            SnapDesigner designer = new SnapDesigner(helper);
            designer.ShowDialog();
        }
        public static void ShowPrintPreview(IDocumentData documentData, IObjectSpace objectSpace) {
            ShowPrintPreview(documentData, objectSpace, null);
        }
        public static void ShowPrintPreview(IDocumentData documentData, IObjectSpace objectSpace, CriteriaOperator inPlaceCriteria) {
            using (SnapDocumentServer server = new SnapDocumentServer()) {
                SnapDocumentHelper helper = new SnapDocumentHelper(documentData, objectSpace, null, inPlaceCriteria);
                helper.LoadDocument(server.Document);
                // Mail Merge
                if (server.Document.DataSources.Count == 1 && server.Options.SnapMailMergeVisualOptions.DataSource == server.Document.DataSources[0].DataSource) {
                    using (MemoryStream stream = new MemoryStream()) {
                        server.SnapMailMerge(stream, SnapDocumentFormat.Snap);
                        stream.Flush();
                        stream.Seek(0, SeekOrigin.Begin);
                        server.LoadDocument(stream, SnapDocumentFormat.Snap);
                    }
                }
                using (PrintingSystem ps = new PrintingSystem()) {
                    PrintableComponentLink link = new PrintableComponentLink(ps);
                    link.Component = server;
                    link.CreateDocument();
                    link.ShowPreviewDialog();
                }
            }
        }
        public void LoadDocument(SnapDocument document) {
            this.document = document;
            if (documentData.Content != null) {
                using (MemoryStream stream = new MemoryStream()) {
                    stream.Write(documentData.Content, 0, documentData.Content.Length);
                    stream.Flush();
                    stream.Seek(0, SeekOrigin.Begin);
                    document.LoadDocument(stream, SnapDocumentFormat.Snap);
                }
            }
            SynchronizeDataSources();
        }

        private void SynchronizeDataSources() {
            List<DataSourceInfo> dataSourcesToRemove = new List<DataSourceInfo>();
            foreach (DataSourceInfo dataSourceInfo in document.DataSources) {
                if (!(dataSourceInfo.DataSource is DataSet)) {
                    bool exists = false;
                    foreach (IDocumentDataSource dataSource in documentData.DataSources) {
                        if (dataSource.Name == dataSourceInfo.DataSourceName) {
                            exists = true;
                            break;
                        }
                    }
                    if (!exists) {
                        dataSourcesToRemove.Add(dataSourceInfo);
                    }
                }
            }
            foreach (DataSourceInfo dataSourceInfo in dataSourcesToRemove) {
                document.DataSources.Remove(dataSourceInfo);
            }
            foreach (IDocumentDataSource dataSource in documentData.DataSources) {
                AddDataSource(dataSource, dataSource == documentData.InplaceDocumentDataSource);
            }
            document.Fields.Update();
        }
        private void AddDataSource(IDocumentDataSource dataSource, bool isInplaceDataSource) {
            CriteriaOperator criteria = isInplaceDataSource ? inPlaceCriteria : null;
            if (!String.IsNullOrEmpty(dataSource.Criteria)) {
                CriteriaOperator dataSourceCriteria = CriteriaEditorHelper.GetCriteriaOperator(dataSource.Criteria, dataSource.DataType, objectSpace);
                if (ReferenceEquals(criteria, null)) {
                    criteria = dataSourceCriteria;
                } else {
                    criteria = new GroupOperator(GroupOperatorType.And, criteria, dataSourceCriteria);
                }
            }
            CollectionSource collectionSource = new CollectionSource(objectSpace, dataSource.DataType, CollectionSourceDataAccessMode.Client, CollectionSourceMode.Proxy);
            collectionSource.Criteria["SnapDataSourceFiler"] = criteria;
            if (document.DataSources[dataSource.Name] != null) {
                document.DataSources[dataSource.Name].DataSource = collectionSource.Collection;
            } else {
                document.DataSources.Add(dataSource.Name, collectionSource.Collection);
            }
        }
        internal void SaveDocument() {
            if (objectSpace.IsNewObject(documentData)) {
                ShowDocumentDataObject();
            }
            using (MemoryStream stream = new MemoryStream()) {
                document.SaveDocument(stream);
                stream.Seek(0, SeekOrigin.Begin);
                byte[] bytes = new byte[stream.Length];
                stream.Read(bytes, 0, bytes.Length);
                documentData.Content = bytes;
                objectSpace.CommitChanges();
            }
        }
        internal void ShowDocumentDataObject() {
            DetailView documentDetailView = application.CreateDetailView(objectSpace, documentData, false);
            documentDetailView.Caption = "Save Document";
            ShowViewParameters svp = new ShowViewParameters(documentDetailView);
            svp.TargetWindow = TargetWindow.NewModalWindow;
            svp.Context = TemplateContext.PopupWindow;
            DialogController dc = application.CreateController<DialogController>();
            dc.SaveOnAccept = false;
            svp.Controllers.Add(dc);
            application.ShowViewStrategy.ShowView(svp, new ShowViewSource(null, null));
            SynchronizeDataSources();
        }
        internal void CreateNewDocument(SnapDocument document) {
            documentData = (IDocumentData)objectSpace.CreateObject(documentData.GetType());
        }
        internal void ShowNewDataSourceDialog(){
            IObjectSpace newDataSourceObjectSpace = objectSpace.CreateNestedObjectSpace();
            Type dataSourceType = XafTypesInfo.Instance.FindTypeInfo(documentData.GetType()).FindMember("DataSources").ListElementType;
            IDocumentDataSource dataSource = (IDocumentDataSource)newDataSourceObjectSpace.CreateObject(dataSourceType);
            DetailView dataSourceDetailView = application.CreateDetailView(newDataSourceObjectSpace, dataSource);
            ShowViewParameters svp = new ShowViewParameters(dataSourceDetailView);
            svp.TargetWindow = TargetWindow.NewModalWindow;
            svp.Context = TemplateContext.PopupWindow;
            DialogController dc = application.CreateController<DialogController>();
            svp.Controllers.Add(dc);
            dc.Accepting += delegate {
                newDataSourceObjectSpace.CommitChanges();
                IDocumentDataSource newDataSource = (IDocumentDataSource)objectSpace.GetObject(dataSource);
                documentData.DataSources.Add(newDataSource);
                AddDataSource(newDataSource, false);
            };
            application.ShowViewStrategy.ShowView(svp, new ShowViewSource(null, null));
        }
        internal bool QueryCanClose() {
            ConfirmationResult result = application.AskConfirmation(ConfirmationType.NeedSaveChanges);
            if (result == ConfirmationResult.Cancel) {
                return false;
            }
            if (result == ConfirmationResult.Yes) {
                SaveDocument();
            }
            return true;
        }
        internal string GetDocumentName() {
            return documentData.DocumentName;
        }
        internal bool Modified {
            get {
                return objectSpace.IsModified;
            }
        }
    }
}
